storagenode/piecestore/endpoint.go

```diff
@@ -552,6 +554,46 @@ func (endpoint *Endpoint) SaveOrder(ctx context.Context, limit *pb.OrderLimit, o
 	}
 }
 
+// Retain keeps only piece ids specified in the request
+func (endpoint *Endpoint) Retain(ctx context.Context, retainReq *pb.RetainRequest) (*pb.RetainResponse, error) {
+	peer, err := identity.PeerIdentityFromContext(ctx)
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	filter, err := bloomfilter.NewFromBytes(retainReq.GetFilter())
+	if err != nil {
+		return nil, Error.Wrap(err)
+	}
+
+	limit := 1000
+	offset := 0
+	piecesRemaining := true
+
+	for piecesRemaining {
+		pieceIDs, err := endpoint.pieceinfo.GetPieceIDs(ctx, peer.ID, retainReq.GetCreationDate(), limit, offset)
+		if err != nil {
+			return nil, Error.Wrap(err)
+		}
+		for _, pieceID := range pieceIDs {
+			if !filter.Contains(pieceID) {
+				if err = endpoint.store.Delete(ctx, peer.ID, pieceID); err != nil {
```

Comment:
hmm well if we delete from the store successfully and the delete from pieceinfo fails, I think that is fine, because on the next round, the store delete will fail (since it won't exist anymore), but that doesn't block the delete from the pieceinfo from occurring and hopefully working again
However, if we delete from the pieceinfo successfully and deleting from the store fails, that is a problem because we won't ever hit this part of the code again. Maybe we should add a todo or something about this?

Reply:
I think we should do the following (if it's easy):
If `endpoint.store.Delete` succeeds, or it has an error saying the piece doesn't exist (because it was already deleted), we follow through with `endpoint.pieceinfo.Delete`.
If `endpoint.store.Delete` results in some other type of error, we should not delete from pieceinfo. That way, we can attempt again on the next round of garbage collection.
We should only do this if we can easily distinguish between a "doesn't exist/was already deleted" error and other errors for `endpoint.store.Delete`

Reply:
Ok - I looked into it and it looks like no error is returned when `endpoint.store.Delete` tries to delete something that is not on the filesystem (see `storage/filestore` `Delete()`). So we can just not delete from pieceinfo if deleting from store results in an error, and otherwise delete from pieceinfo.